### PR TITLE
CDAP-15861 Google Analytics 360 (SaaS source)

### PR DESCRIPTION
**Pull request by anton-kai**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+                        <include>**/Test*.java</include>
+                        <include>**/*Test.java</include>
+                        <include>**/*TestCase.java</include>
+                    </includes>
+                    <excludes>
+                        <exclude>**/*TestRun.java</exclude>
+                    </excludes>
+                </configuration>
+            </plugin>
+<!--            <plugin>-->
```
Comment (mlozbin-cybervisiontech) [inline pom.xml]: why is it commented out?

File: pom.xml [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+                <version>3.3.0</version>
+                <extensions>true</extensions>
+                <configuration>
+                    <instructions>
+                        <Embed-Dependency>*;inline=false;scope=compile</Embed-Dependency>
+                        <Embed-Transitive>true</Embed-Transitive>
+                        <_exportcontents>io.cdap.plugin.ga360.*</_exportcontents>
+                        <Embed-Directory>lib</Embed-Directory>
+                        <!--Only @Plugin classes in the export packages will be included as plugin-->
+                        <_exportcontents>io.cdap.*</_exportcontents>
```
Comment (mlozbin-cybervisiontech) [inline pom.xml]: duplicated entry. Should be removed.

File: pom.xml [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
+                <groupId>io.cdap</groupId>
+                <artifactId>cdap-maven-plugin</artifactId>
+                <version>1.1.0</version>
+                <configuration>
+                    <cdapArtifacts>
+                        <parent>system:cdap-data-pipeline[6.0.0-SNAPSHOT,7.0.0-SNAPSHOT)</parent>
```
Comment (mlozbin-cybervisiontech) [inline pom.xml]: 6.0.0-SNAPSHOT -> 6.1.0-SNAPSHOT

File: src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsBatchSource.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            .map(Schema.Field::getName)
+            .collect(Collectors.toList()));
+
+    batchSourceContext.setInput(Input.of(config.referenceName, new GoogleAnalyticsFormatProvider(config)));
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    FailureCollector failureCollector = pipelineConfigurer.getStageConfigurer().getFailureCollector();
+    config.validate(failureCollector);
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsBatchSource.java]: on this stage all macro-enabled fields may be null. Please check. If it is still true - we should not validate fields with macros on this stage and should do second validation on prepareRun

File: src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsConfig.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+      return Arrays.asList(dimensionsList.split(","));
+    } else {
+      return Collections.emptyList();
+    }
+  }
+
+  @Override
+  public void validate(FailureCollector failureCollector) {
+    super.validate(failureCollector);
+    //TODO extend validation here
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsConfig.java]: why wasn't it extended?

File: src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsRecordReader.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      // Call the batchGet method.
+      GetReportsResponse response = AnalyticsReportingInitializer.initializeAnalyticsReporting()
+          .reports()
+          .batchGet(getReport)
+          .setAccessToken(gaConfig.getAuthorizationToken())
+          .execute();
+
+      currentReport = response.getReports().get(0);
+    } catch (GeneralSecurityException e) {
+      e.printStackTrace(); // TODO handle or throw exception
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsRecordReader.java]: why it hasn't been done?

File: src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsRecordReader.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    } catch (IOException e) {
+      e.printStackTrace(); // TODO handle or throw exception
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsRecordReader.java]: why it hasn't been done?

File: src/main/java/io/cdap/plugin/ga360/source/common/BaseSourceConfig.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+  public String getAuthorizationToken() {
+    return authorizationToken;
+  }
+
+  public String getViewId() {
+    return viewId;
+  }
+
+  public void validate(FailureCollector failureCollector) {
+    if (Strings.isNullOrEmpty(authorizationToken)) {
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/ga360/source/common/BaseSourceConfig.java]: please check how it works if field contains macro

File: src/main/java/io/cdap/plugin/ga360/source/common/BaseSourceConfig.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+      failureCollector
+          .addFailure(String.format("%s must be not empty.", AUTHORIZATION_TOKEN),
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/ga360/source/common/BaseSourceConfig.java]: 1. must be specified
2. correction action may be skipped in this case

File: src/main/java/io/cdap/plugin/ga360/source/common/ReportTransformer.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+import io.cdap.cdap.api.data.schema.Schema;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+public class ReportTransformer {
+
+  public static StructuredRecord transform(Report report, Schema schema) {
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    //TODO write transform logic
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/ga360/source/common/ReportTransformer.java]: ?

File: src/main/java/io/cdap/plugin/ga360/source/common/SchemaBuilder.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Helper class to map Google Analytics Report fields sets to final {@link Schema}.
+ */
+public class SchemaBuilder {
+
+  public static Schema buildSchema(List<String> metrics, List<String> dimensions) {
+    //TODO how to build schema?
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/ga360/source/common/SchemaBuilder.java]: ?

Comment (anton-kai) [reply src/main/java/io/cdap/plugin/ga360/source/common/SchemaBuilder.java]: done

File: src/main/java/io/cdap/plugin/ga360/source/common/ReportTransformer.java (hunk 10 above)
Comment (anton-kai) [reply src/main/java/io/cdap/plugin/ga360/source/common/ReportTransformer.java]: done

File: src/main/java/io/cdap/plugin/ga360/source/common/BaseSourceConfig.java (hunk 9 above)
Comment (anton-kai) [reply src/main/java/io/cdap/plugin/ga360/source/common/BaseSourceConfig.java]: done

File: src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsRecordReader.java (hunk 7 above)
Comment (anton-kai) [reply src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsRecordReader.java]: done

File: pom.xml (hunk 2 above)
Comment (anton-kai) [reply pom.xml]: done

File: pom.xml (hunk 1 above)
Comment (anton-kai) [reply pom.xml]: done

File: pom.xml (hunk 3 above)
Comment (anton-kai) [reply pom.xml]: Done

File: src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsConfig.java (hunk 5 above)
Comment (anton-kai) [reply src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsConfig.java]: done

File: src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsBatchSource.java (hunk 4 above)
Comment (anton-kai) [reply src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsBatchSource.java]: done

File: src/main/java/io/cdap/plugin/ga360/source/common/BaseSourceConfig.java (hunk 8 above)
Comment (anton-kai) [reply src/main/java/io/cdap/plugin/ga360/source/common/BaseSourceConfig.java]: done

File: src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsConfig.java [hunk 12]
```diff
@@ -79,14 +83,20 @@ Schema getSchema() {
...
   public String getStartDate() {
     return startDate;
   }
 
+  @Nullable
   public String getEndDate() {
     return endDate;
   }
 
+  public Boolean isDateRangeFilter() {
```
[the remark this answers is not in the document]
Comment (anton-kai) [reply src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsConfig.java]: removed

File: src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsRecordReader.java [hunk 13]
```diff
@@ -57,20 +58,18 @@ public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptCont
...
+        .batchGet(getReport)
+        .setAccessToken(gaConfig.getAuthorizationToken())
+        .execute();
 
       currentReport = response.getReports().get(0);
     } catch (GeneralSecurityException e) {
-      e.printStackTrace(); // TODO handle or throw exception
-    } catch (IOException e) {
-      e.printStackTrace(); // TODO handle or throw exception
+      throw new IOException(e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (anton-kai) [reply src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsRecordReader.java]: What should  I throw here?

File: pom.xml [hunk 14]
```diff
@@ -401,7 +349,6 @@
                     <instructions>
                         <Embed-Dependency>*;inline=false;scope=compile</Embed-Dependency>
                         <Embed-Transitive>true</Embed-Transitive>
-                        <_exportcontents>io.cdap.plugin.ga360.*</_exportcontents>
                         <Embed-Directory>lib</Embed-Directory>
                         <!--Only @Plugin classes in the export packages will be included as plugin-->
                         <_exportcontents>io.cdap.*</_exportcontents>
```
[the remark this answers is not in the document]
Comment (anton-kai) [reply pom.xml]: It doesn't work without it

[the remark this answers is not in the document]
Comment (mlozbin-cybervisiontech) [reply pom.xml]: it doesn't work with io.cdap.plugin.ga360.*? Why?

File: src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsRecordReader.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (mlozbin-cybervisiontech) [reply src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsRecordReader.java]: IllegalStateException for example

File: pom.xml (hunk 14 above)
[the remark this answers is not in the document]
Comment (anton-kai) [reply pom.xml]: It throws ClassNotFound during MapReduce job

File: src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsRecordReader.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (anton-kai) [reply src/main/java/io/cdap/plugin/ga360/source/batch/GoogleAnalyticsRecordReader.java]: fixed

File: pom.xml (hunk 14 above)
[the remark this answers is not in the document]
Comment (mlozbin-cybervisiontech) [reply pom.xml]: so please find a way to fix it

[the remark this answers is not in the document]
Comment (anton-kai) [reply pom.xml]: fixed

**Pull request merged**